### PR TITLE
🌱 OptionValues helper

### LINKED ISSUE #570: Minor refactor of ExtraConfig methods -- pkg/utils/vsphere
**Please describe the solution you would like.**

https://github.com/vmware-tanzu/vm-operator/pull/569#issuecomment-2155584228
In pkg/utils/vsphere, I would:

```
func OptionValuesFromList(in []vimtypes.BaseOptionValue) OptionValues {
    if len(in) == 0 {
        return nil
    }

    out := make(OptionValues{}, len(in))

    for i := range in {
        switch tval := in[i].Value.(type) {
        case string:
            out[in[i].Key] = tval
        default:
            out[in[i].Key] = fmt.Sprintf("%v", tval)
        }
    }

    return out
}

type OptionValues map[string]string

func (ov OptionValues) ToList() []vimtypes.BaseOptionValue {}

func (ov OptionValues) Merge(in OptionValues) {}

func (ov OptionValues) MergeList(in []vimtypes.BaseOptionValue) {}
```
Something like the above. A very simple and efficient way to intake, manipulate, and reuse ExtraConfig or anything else that is a list of option values.

**Is there anything else you would like to add?**

N/A

**Please tell us about your environment.**
N/A

### PR DESCRIPTION
**Pull request by akutz**
**What does this PR do, and why is it needed?**

This patch introduces the type OptionValues in "pkg/util". This new type simplifies working with []vimtypes.BaseOptionValue properties such as ExtraConfig.

@dougm , it occurs to me after writing this ... should this just be in GoVmomi?

**Which issue(s) is/are addressed by this PR?** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #570

**Are there any special notes for your reviewer**:

Please note there are _two_ commits for this PR:

1. The actual business logic.
2. A patch where the files affected from the previous patch have their `pkg/util` packages imported according to the correct rules as `pkgutil`. 

This enables reviewers to focus on the business logic patch separately, making it easier to see what actually changed.

**Please add a release note if necessary**:

```release-note
Introduce pkg/util.OptionValues for working with properties like ExtraConfig
```

### CODE REVIEW

**Comment by dougm** [comment]
> it occurs to me after writing this ... should this just be in GoVmomi?

I was thinking that too, along with some of the recent resize work that has gone in (which could be used in simulator at least). We could start a sheet/list of vmop helpers that would be generally useful if they were moved into govmomi. And we could use that to discuss which package(s) such helpers would land in govmomi.

File: pkg/util/option_values.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+// AppendList combines the provided input with the current OptionValues data,
+// ignoring any keys from the input data that already exist in the existing
+// data.
+func (ov OptionValues) AppendList(in []vimtypes.BaseOptionValue) {
+	ov.mergeList(in, false)
+}
+
+// Merge combines the provided input with the current OptionValues data. Any
+// matching keys from the input data will overwrite keys in the existing data.
+func (ov OptionValues) Merge(in OptionValues) {
```
Comment (aruneshpa) [inline pkg/util/option_values.go:115]: I don't see this one being called anywhere as well.  If this is just for future proofing, I would vote for adding it when it's needed.  Sorry if I missed where it is referenced in the diff.

File: pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go [hunk 2]
```diff
@@ -253,8 +253,12 @@ func GetCloudInitGuestInfoCustSpec(
 		extraConfig[constants.CloudInitGuestInfoUserdataEncoding] = "gzip+base64"
 	}
 
-	configSpec := &vimtypes.VirtualMachineConfigSpec{}
-	configSpec.ExtraConfig = util.MergeExtraConfig(config.ExtraConfig, extraConfig)
+	newExtraConfig := pkgutil.OptionValuesFromList(config.ExtraConfig)
+
+	configSpec := &vimtypes.VirtualMachineConfigSpec{
+		ExtraConfig: newExtraConfig.Diff(extraConfig).ToList(),
```
[the remark this answers is not in the document]
Comment (akutz) [reply pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go:259]: No, it is not merging as it does not impact the *existing* `OptionValues` object. It returns a difference of what is required to update the original based on the new values.

File: pkg/util/resize/configspec.go [hunk 3]
```diff
@@ -265,8 +265,8 @@ func compareExtraConfig(
 		return
 	}
 
-	extraConfig := util.ExtraConfigToMap(cs.ExtraConfig)
-	outCS.ExtraConfig = util.MergeExtraConfig(ci.ExtraConfig, extraConfig)
+	extraConfig := pkgutil.OptionValuesFromList(ci.ExtraConfig)
+	outCS.ExtraConfig = extraConfig.DiffListToList(cs.ExtraConfig)
```
[the remark this answers is not in the document]
Comment (akutz) [reply pkg/util/resize/configspec.go:269]: Nope, because of the ordering. Please note the comments on the functions. The `ToList` variants specifically maintain insertion order at the request of @bryanv to make ConfigSpec comparisons easier for resize.

File: pkg/util/option_values.go (hunk 1 above)
Comment (akutz) [reply pkg/util/option_values.go:115]: The old function was called `MergeExtraConfig`, but it was improperly named. It should have been called `DiffExtraConfig`. I am keeping the `Merge` here in case someone _does_ need it in the future to ensure it is written the correct way.

File: pkg/util/option_values.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	}
+	if len(out) == 0 {
+		return nil
+	}
+	return out
+}
+
+// DiffList returns the difference between the existing data and provided input.
+// Please note the output does not maintain the order of the input.
+func (ov OptionValues) DiffList(in []vimtypes.BaseOptionValue) OptionValues {
```
[the remark this answers is not in the document]
Comment (akutz) [reply pkg/util/option_values.go:63]: It is for consistency with the other functions, Append and Merge. Both of them have List options. This is to ensure folks that _want_ to use them have correctly written functions to use.

**Comment by akutz** [comment]
@aruneshpa If you are good with our discussion, did you want to approve this?

File: pkg/providers/vsphere/session/session_vm_update.go [hunk 5]
```diff
@@ -326,8 +326,7 @@ func UpdateConfigSpecExtraConfig(
 		// with the class config spec extra config (ie) class config spec extra
 		// config keys takes precedence over the desired config spec extra
 		// config keys.
-		combinedExtraConfig := util.AppendNewExtraConfigValues(classConfigSpec.ExtraConfig, extraConfig)
-		extraConfig = util.ExtraConfigToMap(combinedExtraConfig)
+		extraConfig.Append(pkgutil.OptionValuesFromList(classConfigSpec.ExtraConfig))
```
Comment (sreyasn) [inline pkg/providers/vsphere/session/session_vm_update.go:329]: can we use AppendList() here ?

File: pkg/providers/vsphere/vmprovider_vm.go [hunk 6]
```diff
@@ -1033,8 +1033,10 @@ func (vs *vSphereVMProvider) vmCreateGenConfigSpecExtraConfig(
 		}
 	}
 
-	// The ConfigSpec's current ExtraConfig values (that came from the class) take precedence over what was set here.
-	createArgs.ConfigSpec.ExtraConfig = util.AppendNewExtraConfigValues(createArgs.ConfigSpec.ExtraConfig, ecMap)
+	// The ConfigSpec's current ExtraConfig values (that came from the class)
+	// take precedence over what was set here.
+	newExtraConfig := pkgutil.OptionValues(ecMap)
+	newExtraConfig.MergeList(createArgs.ConfigSpec.ExtraConfig)
```
Comment (sreyasn) [inline pkg/providers/vsphere/vmprovider_vm.go:1039]: I see a behavior change from what happened with AppendNewExtraConfigValues(). The earlier method only added new fields from _ecMap_ that didn't exist in _createArgs.ConfigSpec.ExtraConfig_. 

Might need to do something like IIUC
```
createArgsEC := pkgutil.OptionValuesFromList(createArgs.ConfigSpec.ExtraConfig)
newEc := pkgutil.OptionValues(ecMap)
createArgsEc.Append(newEc)
```

File: pkg/util/resize/configspec.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sreyasn) [reply pkg/util/resize/configspec.go:269]: curious: Are we sure that by insertion ordering here, we also maintain the order that we get back from the configInfo when fetched from the VM?.

File: pkg/util/option_values.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+// Please note the output does not maintain the order of the input.
+func (ov OptionValues) Diff(in OptionValues) OptionValues {
+	out := OptionValues{}
+	for k, newVal := range in {
+		if oldVal, ok := ov[k]; !ok || newVal != oldVal {
+			out[k] = newVal
+		}
+	}
+	if len(out) == 0 {
+		return nil
```
Comment (bryanv) [inline pkg/util/option_values.go:56]: Depending on how these calls are all chain, this does worry me a little that we can unexpectedly NPE, like if somebody would then call a func that ends up in `merge()` or `mergeList()`

File: pkg/util/resize/configspec.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bryanv) [reply pkg/util/resize/configspec.go:269]: There are some situations for writing simpler test assertions a deterministic ordering is helpful, but consider that ask secondary to whatever we want to have clearer non-test code.

File: pkg/util/option_values.go (hunk 7 above)
Comment (akutz) [reply pkg/util/option_values.go:56]: How so? We allow `in` to be `nil` without error, and the functions that could NPE do not return anything, such as `Merge` and `MergeList`. These specifically do not return and cannot be used to chain calls.

Comment (bryanv) [reply pkg/util/option_values.go:56]: Let me rephrase a little bit - with the type alias you still need need to know the underlying type is nullable. I'm worried that something like below is too easy to write or end up with:

```
in := []vimtypes.BaseOptionValue{&vimtypes.OptionValue{Key: "foo", Value: "one"}}
more := []vimtypes.BaseOptionValue{&vimtypes.OptionValue{Key: "foo1", Value: "two"}}
x := pkgutil.OptionValuesFromList(in).DiffList(in)
....
x.MergeList(more)
```

Thinking back to a bug that was introduced with the switch to either the `slices` or `maps` package that an empty in param return nil.

**Review by sreyasn — APPROVED** [review]

**Review by bryanv — APPROVED** [review]
